_posts/2018-12-05-exploit-model.md

```diff
@@ -0,0 +1,153 @@
+
+>The backing memory is then reclaimed and handed to another (potentially more privileged) process.
+
+This may perhaps lead to possibilities to obtain more attacker capabilities in the gvisor instance - maybe even code exec? We'll discuss this next.
+
+
+## Exploring differences in LibOS security model
+
+Let's dive in a little deeper into a scenario where hypothetically, an attacker obtains code exec capabilities in gvisor. So assuming that we are able to exploit and get code exec in gvisor, what protections do we get in terms of isolation? 
+
+Syscalls are the way that an application interfaces with gvisor, and we can view it as an estimation of the attack surface[^1] from the application to gvisor. We've seen here, that this exposure of 300+ syscalls consists of `shmctl`, which allowed the application process to trigger a bug in gvisor.
+
+So assuming that we've obtained code exec in gvisor, what is our exposure. We looked at the [seccomp policy of gvisor](https://github.com/google/gvisor/blob/0091db9cbddb6c9fb4c96fbde980780c98006eda/runsc/boot/filter/config.go), and it is assuring to see that the seccomp policy only allows 83 or the 300+ syscalls available. Therefore, we note that given an exploit in the gvisor kernel, the attack surface is much reduced compared to regular containers.
+
+```
++--------+              +---------+
+|  p1a   |              |   p2a   |                      APP
++--------+              +---------+
+<<< 300 syscalls  >>>   <<< 300 syscalls  >>>
++-------------------+   +-------------------+
+|      gvisor       |   |      gvisor       |            GVISOR
+|   (code exec)     |   |                   |            
++-------------------+   +-------------------+
+<<<  83 syscalls  >>>   <<<  83 syscalls  >>>     <--- Attack surface 
++--------------------------------------------------+
+|                    kernel                        |     KERNEL
++--------------------------------------------------+
+```
+
+Based on this, in order to break isolation between gvisor instances, an attacker would need to obtain code exec or something similar in the kernel via the 83 syscalls allowed.
+
+### Another model: LibOS with Nabla
+
+Let's compare the Nabla containers model with that of gvisor. The components are very similar, gvisor is a userspace unikernel, and Nabla has a Library OS (LibOS). This LibOS does a similar function - implementing the functionality of the kernel in userspace and only calling the kernel when necessary. 
+
+We note that right now, we don't distinguish between the application and the LibOS, and we treat the LibOS as untrusted. For the application and LibOS, we have a seccomp policy that only allows 7 syscalls, with only two file descriptions (a block device and tap device).
```

Comment:
So: neither LibOS nor gVisor is part of the TCB?

Reply:
In the nabla case, the LibOS is not part of the host TCB. For gVisor, it is considered to be semi-trusted, it is another security boundary.